File: matching.py
"""
匹配引擎
规则初筛（纯代码） + MiMo 精排
"""
import json
from datetime import datetime, timedelta
from models import get_db
from mimo_client import match_candidates, generate_icebreaker


def rule_filter(user_id: int, scene: str) -> list:
    """
    规则初筛：从数据库中筛选满足基本条件的候选人
    规则：同场景 + 时间差≤30min + 排除自己
    """
    conn = get_db()

    # 获取当前用户画像
    my_profile = conn.execute(
        'SELECT * FROM profiles WHERE user_id = ? AND scene = ?',
        (user_id, scene)
    ).fetchone()

    if not my_profile:
        conn.close()
        return []

    # 初筛：同场景 + 非自己 + 有画像
    candidates = conn.execute('''
        SELECT p.*, u.nickname, u.department, u.avatar_url
        FROM profiles p
        JOIN users u ON p.user_id = u.id
        WHERE p.scene = ? AND p.user_id != ?
    ''', (scene, user_id)).fetchall()

    results = []
    for c in candidates:
        score = 0

        # 时间匹配 (30分)
        if my_profile['time_pref'] and c['time_pref']:
            my_time = _parse_time(my_profile['time_pref'])
            c_time = _parse_time(c['time_pref'])
            if my_time and c_time:
                diff = abs((my_time - c_time).total_seconds()) / 60
                if diff <= 15:
                    score += 30
                elif diff <= 30:
                    score += 15
                else:
                    continue  # 时间差>30min，直接排除
            else:
                score += 15  # 无法解析，给半分

        # 地点/路线匹配 (25分)
        if my_profile['location_pref'] and c['location_pref']:
            if my_profile['location_pref'] == c['location_pref']:
                score += 25
            else:
                score += 10

        # 口味/预算匹配 (15分)
        my_taste = set(json.loads(my_profile['taste_pref'] or '[]'))
        c_taste = set(json.loads(c['taste_pref'] or '[]'))
        if my_taste and c_taste:
            overlap = my_taste & c_taste
            if overlap == my_taste:
                score += 15
            elif overlap:
                score += 8

        # 兴趣标签匹配 (15分)
        my_interests = set(json.loads(my_profile['interests'] or '[]'))
        c_interests = set(json.loads(c['interests'] or '[]'))
        if my_interests and c_interests:
            common = my_interests & c_interests
            total = my_interests | c_interests
            if total:
                score += int(len(common) / len(total) * 15)

        # 社交偏好匹配 (15分)
        if my_profile['social_pref'] and c['social_pref']:
            if my_profile['social_pref'] == c['social_pref']:
                score += 15
            else:
                score += 8

        if score > 0:
            results.append({
                'user_id': c['user_id'],
                'nickname': c['nickname'],
                'department': c['department'],
                'avatar_url': c['avatar_url'],
                'rule_score': score,
                'profile': dict(c)
            })

    conn.close()

    # 按规则分数排序，取 Top10
    results.sort(key=lambda x: x['rule_score'], reverse=True)
    return results[:10]


def do_match(user_id: int, scene: str) -> list:
    """
    完整匹配流程：规则初筛 → MiMo 精排 → 返回 Top3
    """
    # 步骤1：规则初筛
    candidates = rule_filter(user_id, scene)

    if not candidates:
        return []

    # 步骤2：MiMo 精排
    conn = get_db()
    my_profile = conn.execute(
        'SELECT * FROM profiles WHERE user_id = ? AND scene = ?',
        (user_id, scene)
    ).fetchone()

    if my_profile:
        profile_dict = dict(my_profile)
        mimo_results = match_candidates(profile_dict, candidates, scene)

        # 合并规则分数和 MiMo 分数
        for r in mimo_results:
            for c in candidates:
                if c['user_id'] == int(r.get('candidate_id', 0)):
                    r['rule_score'] = c['rule_score']
                    r['nickname'] = c['nickname']
                    r['avatar_url'] = c['avatar_url']

        # 按 MiMo 分数排序，取 Top3
        mimo_results.sort(key=lambda x: x.get('score', 0), reverse=True)
        conn.close()
        return mimo_results[:3]

    conn.close()
    return candidates[:3]


def _parse_time(time_str: str):
    """尝试解析时间字符串"""
    for fmt in ['%H:%M', '%H:%M:%S']:
        try:
            return datetime.strptime(time_str, fmt)
        except ValueError:
            continue
    return None

File: services/matching.js
/**
 * 匹配引擎
 * 规则初筛（纯代码） + MiMo 精排
 */
const { pool } = require('../db');
const { matchCandidates, generateIcebreaker } = require('./mimo');

/**
 * 规则初筛：同场景 + 时间差≤30min + 排除自己，按规则分数排序取 Top10
 */
async function ruleFilter(userId, scene) {
  const [myRows] = await pool.query(
    'SELECT * FROM profiles WHERE user_id = ? AND scene = ?',
    [userId, scene]
  );
  if (myRows.length === 0) return [];
  const myProfile = myRows[0];

  const [candidates] = await pool.query(`
    SELECT p.*, u.nickname, u.department, u.avatar_url
    FROM profiles p
    JOIN users u ON p.user_id = u.id
    WHERE p.scene = ? AND p.user_id != ?
  `, [scene, userId]);

  const results = [];
  for (const c of candidates) {
    let score = 0;

    // 时间匹配 (30分)
    if (myProfile.time_pref && c.time_pref) {
      const diff = timeDiffMinutes(myProfile.time_pref, c.time_pref);
      if (diff !== null) {
        if (diff <= 15) score += 30;
        else if (diff <= 30) score += 15;
        else continue; // 时间差>30min，排除
      } else {
        score += 15;
      }
    }

    // 地点/路线匹配 (25分)
    if (myProfile.location_pref && c.location_pref) {
      if (myProfile.location_pref === c.location_pref) score += 25;
      else score += 10;
    }

    // 口味匹配 (15分)
    const myTaste = parseJSON(myProfile.taste_pref, []);
    const cTaste = parseJSON(c.taste_pref, []);
    if (myTaste.length && cTaste.length) {
      const overlap = myTaste.filter(t => cTaste.includes(t));
      if (overlap.length === myTaste.length) score += 15;
      else if (overlap.length > 0) score += 8;
    }

    // 兴趣标签匹配 (15分)
    const myInterests = parseJSON(myProfile.interests, []);
    const cInterests = parseJSON(c.interests, []);
    if (myInterests.length && cInterests.length) {
      const common = myInterests.filter(t => cInterests.includes(t));
      const total = [...new Set([...myInterests, ...cInterests])];
      if (total.length > 0) score += Math.floor(common.length / total.length * 15);
    }

    // 社交偏好匹配 (15分)
    if (myProfile.social_pref && c.social_pref) {
      if (myProfile.social_pref === c.social_pref) score += 15;
      else score += 8;
    }

    if (score > 0) {
      results.push({
        user_id: c.user_id,
        nickname: c.nickname,
        department: c.department,
        avatar_url: c.avatar_url,
        rule_score: score,
        profile: c,
      });
    }
  }

  results.sort((a, b) => b.rule_score - a.rule_score);
  return results.slice(0, 10);
}

/**
 * 完整匹配流程：规则初筛 → MiMo 精排 → 返回 Top3
 */
async function doMatch(userId, scene) {
  const candidates = await ruleFilter(userId, scene);
  if (candidates.length === 0) return [];

  const [myRows] = await pool.query(
    'SELECT * FROM profiles WHERE user_id = ? AND scene = ?',
    [userId, scene]
  );
  if (myRows.length === 0) return candidates.slice(0, 3);

  const mimoResults = await matchCandidates(myRows[0], candidates, scene);

  // 合并规则分数和 MiMo 分数
  for (const r of mimoResults) {
    const matched = candidates.find(c => c.user_id === Number(r.candidate_id));
    if (matched) {
      r.rule_score = matched.rule_score;
      r.nickname = matched.nickname;
      r.avatar_url = matched.avatar_url;
    }
  }

  mimoResults.sort((a, b) => (b.score || 0) - (a.score || 0));
  return mimoResults.slice(0, 3);
}

// --- 工具函数 ---

function timeDiffMinutes(t1, t2) {
  const m1 = parseTime(t1);
  const m2 = parseTime(t2);
  if (m1 === null || m2 === null) return null;
  return Math.abs(m1 - m2);
}

function parseTime(str) {
  if (!str) return null;
  const parts = str.split(':');
  if (parts.length < 2) return null;
  return parseInt(parts[0]) * 60 + parseInt(parts[1]);
}

function parseJSON(val, fallback) {
  if (!val) return fallback;
  if (Array.isArray(val)) return val;
  try { return JSON.parse(val); } catch { return fallback; }
}

module.exports = { ruleFilter, doMatch };
